### ISSUE TITLE
Cannot upload ISO, connection reset by peer

### ISSUE THREAD

**Issue by davidpanic** [kubevirt/containerized-data-importer#2608]
**What happened**:

```
$ virtctl image-upload pvc windows-server-oct-2022 --size 6Gi --image-path=en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso --insecure --wait-secs=240
PVC default/windows-server-oct-2022 not found 
PersistentVolumeClaim default/windows-server-oct-2022 created
Waiting for PVC windows-server-oct-2022 upload pod to be ready...
Pod now ready
Uploading data to https://192.168.250.100:31001

 139.62 MiB / 4.84 GiB [=====>---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------]   2.82% 1s

You are using a client virtctl version that is different from the KubeVirt version running in the cluster
Client Version: v0.59.0-alpha.1
Server Version: v0.58.0
Post "https://192.168.250.100:31001/v1alpha1/upload-async": write tcp 192.168.249.140:57350->192.168.250.100:31001: write: connection reset by peer
```

**What you expected to happen**:
I expected to be able to upload the image.

**How to reproduce it (as minimally and precisely as possible)**:
Uhh.. not sure, run the command I guess?

**Additional context**:

cdi-apiserver logs:
```
I0224 12:51:57.325594       1 apiserver.go:83] Note: increase the -v level in the api deployment for more detailed logging, eg. -v=2 or -v=3
W0224 12:51:57.325718       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0224 12:51:57.527656       1 certwatcher.go:125] Updated current TLS certificate
I0224 12:51:57.531991       1 certwatcher.go:81] Starting certificate watcher
2023/02/24 12:52:40 http: TLS handshake error from 10.244.0.0:13132: EOF
2023/02/24 12:56:50 http: TLS handshake error from 10.244.2.0:1808: EOF
2023/02/24 12:56:50 http: TLS handshake error from 10.244.2.0:37864: EOF
2023/02/24 12:56:55 http: TLS handshake error from 10.244.2.0:26046: EOF
2023/02/24 12:56:55 http: TLS handshake error from 10.244.2.0:39700: EOF
2023/02/24 12:56:55 http: TLS handshake error from 10.244.1.1:58847: EOF
2023/02/24 13:07:40 http: TLS handshake error from 10.244.2.0:46828: EOF
2023/02/24 13:07:40 http: TLS handshake error from 10.244.2.0:30332: EOF
```

```
$ kubectl -n kubevirt-hyperconverged get pod
NAME                                                   READY   STATUS    RESTARTS       AGE
bridge-marker-5c7ft                                    1/1     Running   2 (15m ago)    95m
bridge-marker-6dh2c                                    1/1     Running   0              95m
bridge-marker-hrch6                                    1/1     Running   0              95m
cdi-apiserver-5544b977d9-5g4q8                         1/1     Running   0              20m
cdi-deployment-6d4bdf7654-n47nm                        1/1     Running   1 (18m ago)    20m
cdi-operator-6ffc6bdd-5tqkl                            1/1     Running   0              20m
cdi-uploadproxy-cff544d79-jtdl6                        1/1     Running   0              20m
cluster-network-addons-operator-5bd9f8bc49-ll5mk       2/2     Running   0              78m
hco-operator-9bdcf947b-g94jz                           1/1     Running   1 (74m ago)    78m
hco-webhook-7f89ddfb44-nzjfb                           1/1     Running   0              78m
hostpath-provisioner-operator-84cb5dc778-7xngx         1/1     Running   14 (18m ago)   78m
hyperconverged-cluster-cli-download-6969cb5f8f-wz44w   1/1     Running   0              78m
kube-cni-linux-bridge-plugin-9vthl                     1/1     Running   2 (15m ago)    95m
kube-cni-linux-bridge-plugin-ft2kv                     1/1     Running   0              95m
kube-cni-linux-bridge-plugin-l7642                     1/1     Running   0              95m
kubemacpool-cert-manager-855b5c554d-rcsf5              1/1     Running   0              78m
kubemacpool-mac-controller-manager-bfc4b64b7-fsfwn     2/2     Running   0              78m
multus-m5lrx                                           1/1     Running   0              95m
multus-qqfw8                                           1/1     Running   2 (15m ago)    95m
multus-x42jc                                           1/1     Running   0              95m
ssp-operator-7c6748bccb-46bkw                          1/1     Running   0              78m
tekton-tasks-operator-5db8c7b474-z5tpk                 1/1     Running   0              78m
virt-api-6b45748b5-zjksv                               1/1     Running   1 (37m ago)    78m
virt-controller-87cfbb46f-q4fh7                        1/1     Running   4 (18m ago)    78m
virt-exportproxy-dbbfbf485-dcg9b                       1/1     Running   0              94m
virt-handler-cg6w9                                     1/1     Running   0              94m
virt-handler-gdkc8                                     1/1     Running   0              94m
virt-handler-mptlm                                     1/1     Running   1 (15m ago)    70m
virt-operator-7b8fbb9786-w7mzn                         1/1     Running   0              78m
virt-operator-7b8fbb9786-wxbsx                         1/1     Running   1 (74m ago)    130m
```

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`): `quay.io/kubevirt/cdi-controller@sha256:c0264d9f06dea74803d95beb7c90c46bb47ae127ccf85bf26c51acd8877d7e22 (app.kubernetes.io/version: 1.8.0)`
- Kubernetes version (use `kubectl version`): `v1.26.1`
- DV specification: N/A
- Cloud provider or hardware configuration: Bare metal, three nodes, deployed using kubeadm, apiserver balanced with kube-vip
- OS (e.g. from /etc/os-release): `Debian GNU/Linux 11 (bullseye)`
- Kernel (e.g. `uname -a`): `5.10.0-21-amd64 #1 SMP Debian 5.10.162-1 (2023-01-21)`
- Install tools: Installed with the kubevirt-hyperconverged operator using the marketplace.
- Others: N/A

**Comment by davidpanic** [comment]
I was able to get the ISO onto the cluster by making it download it over HTTP. I started up a local web server and used the following manifest:

<details>
<summary>Mainfest (click to expand)</summary>

```yaml
apiVersion: cdi.kubevirt.io/v1alpha1
kind: DataVolume
metadata:
  name: windows-server-oct-2022
  namespace: default
spec:
  source:
    http:
      url: http://192.168.a.b:8080/en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso
  pvc:
    accessModes:
      - ReadWriteOnce
    resources:
      requests:
        storage: 6Gi
```
</details>

That is not a solution to the issue, but rather a workaround.

**Comment by awels** [comment]
Looking at your command, I suspect the virtual image size of image is larger than 6Gi. You can run `qemu-img info en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso` to get the virtual image size. I suspect that is more like 60Gi instead of the 6Gi. If you could give the size slightly more than then virtual image size I think it will work better.

**Comment by davidpanic** [comment]
```
$ qemu-img info en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso
image: en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso
file format: raw
virtual size: 4.84 GiB (5193527296 bytes)
disk size: 4.84 GiB
```

**Comment by awels** [comment]
Okay I was wrong, 6Gi should be plenty. Do you happen to have the logs of the upload-server pod? It should be running in the namespace you are uploading to. If the container crashed and a new one is running (which is likely given the connection reset by peer) you can add `-p` to get the logs of the previous container.

**Comment by davidpanic** [comment]
I don't have any pods whose name contains `upload-server`, however I do have a `cdi-uploadproxy` pod with 0 restarts.

```
$ kubectl get pod --all-namespaces | head -n 1 && kubectl get pod --all-namespaces | grep upload
NAMESPACE                 NAME                                                   READY   STATUS      RESTARTS          AGE
kubevirt-hyperconverged   cdi-uploadproxy-cff544d79-x66nz                        1/1     Running     0                 2d2h
```

The logs don't contain much of anything but I did notice the pod name is different now so it was probably recreated. I think I remember looking at the logs when it was working and it wasn't much different than the below, that's why I only added the `cdi-apiserver` logs in the initial issue report.

<details>
<summary>cdi-uploadproxy pod logs (click to expand)</summary>

```
$ kubectl -n kubevirt-hyperconverged logs cdi-uploadproxy-cff544d79-x66nz
I0227 12:50:56.712986       1 uploadproxy.go:64] Note: increase the -v level in the api deployment for more detailed logging, eg. -v=2 or -v=3
W0227 12:50:56.713968       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0227 12:50:56.833704       1 certwatcher.go:125] Updated current TLS certificate
I0227 12:50:56.834021       1 certwatcher.go:81] Starting certificate watcher
W0227 12:52:46.084244       1 reflector.go:442] pkg/client/informers/externalversions/factory.go:118: watch of *v1beta1.CDIConfig ended with: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeeding
I0227 13:50:43.900277       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.crt": REMOVE
I0227 13:50:43.900596       1 certwatcher.go:125] Updated current TLS certificate
I0227 13:50:43.900656       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.key": REMOVE
I0227 13:50:43.900911       1 certwatcher.go:125] Updated current TLS certificate
I0228 01:50:36.043207       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.key": REMOVE
I0228 01:50:36.043652       1 certwatcher.go:125] Updated current TLS certificate
I0228 01:50:36.043724       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.crt": REMOVE
I0228 01:50:36.044023       1 certwatcher.go:125] Updated current TLS certificate
I0228 13:51:29.036021       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.crt": REMOVE
I0228 13:51:29.036386       1 certwatcher.go:125] Updated current TLS certificate
I0228 13:51:29.036438       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.key": REMOVE
I0228 13:51:29.036729       1 certwatcher.go:125] Updated current TLS certificate
I0301 01:51:25.048155       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.key": REMOVE
I0301 01:51:25.048608       1 certwatcher.go:125] Updated current TLS certificate
I0301 01:51:25.048724       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.crt": REMOVE
I0301 01:51:25.049036       1 certwatcher.go:125] Updated current TLS certificate
I0301 13:51:26.965411       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.key": REMOVE
I0301 13:51:26.965724       1 certwatcher.go:125] Updated current TLS certificate
I0301 13:51:26.965763       1 certwatcher.go:136] certificate eventevent"/var/run/certs/cdi-uploadproxy-server-cert/tls.crt": REMOVE
I0301 13:51:26.965980       1 certwatcher.go:125] Updated current TLS certificate
```
</details>

What's even weirder is that I can successfully upload the image now, with the exact same command. The only thing that comes to mind that is different is that I did disable apparmor on the nodes and increased `nofile` limits in `/etc/security/limits.conf` in between, because some other stuff wasn't compatible.

```
$ virtctl image-upload pvc windows-server-oct-2022 --size 6Gi --image-path=en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso --insecure --wait-secs=240
PVC default/windows-server-oct-2022 not found 
PersistentVolumeClaim default/windows-server-oct-2022 created
Waiting for PVC windows-server-oct-2022 upload pod to be ready...
Pod now ready
Uploading data to https://192.168.250.100:31001

 4.84 GiB / 4.84 GiB [========================================================================================================================================================================================================] 100.00% 1m34s

Uploading data completed successfully, waiting for processing to complete, you can hit ctrl-c without interrupting the progress
Processing completed successfully
Uploading en-us_windows_server_2022_updated_oct_2022_x64_dvd_c5b651c9.iso completed successfully
```

Feel free to close this now since it's working as intended, or ask me for more info if you think you can get anything useful from my cluster or help me reproduce it.

**Comment by awels** [comment]
Let me give you some background on how the upload works. I think apparmor is likely the culprit of why it didn't work. But essentially we don't want to use the k8s api server as a proxy so we have a separate upload-proxy that is running in the cdi installation namespace (in your case the kubevirt-hyperconverged namespace).

Then in the target namespace a pod is created that has `upload-server` in the name. When you run the virtctl command, it connects to the proxy, and the proxy forwards the data to the upload server in the target namespace. This gives us a stable base URL and we can figure out the path on the proxy based on the target namespace and target pvc. When you see `Waiting for PVC windows-server-oct-2022 upload pod to be ready...` that is waiting for the upload server pod to be ready to accept the data.

In short virtctl connects to the upload-proxy, and the upload-proxy connects to the upload-server without ever touching the k8s api server. If either the proxy connection or the upload server connection is disrupted the upload will fail in the way you saw. It is most likely the upload-server pod crashed and that disrupted the connection. Why it crashed should be in the logs.

Closing per your request. Hopefully this kind of explained the process and it potentially failed.

**Comment by caijian76** [comment]
I meet the same error .
root@k8s1:~# virtctl image-upload pvc win101-iso --size=6Gi --image-path=Win10_22H2_China_GGK_Chinese_Simplified_x64.iso  --uploadproxy-url https://cdi-uploadproxy.cdi --insecure --storage-class rook-cephfs 
PVC default/win101-iso not found 
PersistentVolumeClaim default/win101-iso created
Waiting for PVC win101-iso upload pod to be ready...
Pod now ready
Uploading data to https://cdi-uploadproxy.cdi

 2.17 GiB / 5.30 GiB [======================================================>-----------------------------------------------------------------------------]  40.95% 26s

unexpected return value 502,

**Comment by davidpanic** [comment]
I've been debugging this further and enabling jumbo frames (MTU 9000) on all my node network interfaces made things work better for me.

**Comment by tclxzhengyang-li** [comment]
Is there a solution to this problem? I don't have  upload-server pod， only have upload-proxy pod
